src/runtime_src/hip/api/hip_stream.cpp

```diff
@@ -2,35 +2,78 @@
 // Copyright (C) 2023-2024 Advanced Micro Device, Inc. All rights reserved.
 
 #include "hip/core/common.h"
+#include "hip/core/event.h"
 #include "hip/core/stream.h"
 
 namespace xrt::core::hip {
-static hipStream_t
+// In Hip, based on flags we can create default or non blocking streams.
+// If application doesn't explicitly specify stream we use default stream
+// for such operations. The default stream also has two modes legacy or per thread.
+// Legacy default stream is also called as NULL stream. Null stream waits on all explictly
+// created default streams in the same context when an operation is enqueued and explicitly
+// created default streams wait on null stream in that context.
+// Per thread stream is also default stream but is created per thread per context and waits
+// on null stream of that context.
+
+static stream_handle
 hip_stream_create_with_flags(unsigned int flags)
 {
-  throw std::runtime_error("Not implemented");
+  throw_invalid_value_if(flags != hipStreamDefault && flags != hipStreamNonBlocking,
+                         "Invalid flags passed for stream creation");
+  auto hip_ctx = get_current_context();
+  throw_context_destroyed_if(!hip_ctx, "context is destroyed, no active context");
+
+  return insert_in_map(stream_cache, std::make_shared<stream>(hip_ctx, flags));
 }
 
 static void
 hip_stream_destroy(hipStream_t stream)
 {
   throw_invalid_handle_if(!stream, "stream is nullptr");
+  throw_invalid_resource_if(stream == hipStreamPerThread, "Stream per thread can't be destroyed");
 
-  throw std::runtime_error("Not implemented");
+  stream_cache.remove(stream);
 }
 
 static void
 hip_stream_synchronize(hipStream_t stream)
 {
-  throw std::runtime_error("Not implemented");
+  auto hip_stream = get_stream(stream);
+  throw_invalid_handle_if(!hip_stream, "stream is invalid");
+  hip_stream->synchronize();
 }
 
 static void
-hip_stream_wait_event(hipStream_t stream, hipEvent_t event, unsigned int flags)
+hip_stream_wait_event(hipStream_t stream, hipEvent_t ev, unsigned int flags)
 {
-  throw_invalid_handle_if(!event, "event is nullptr");
+  throw_invalid_handle_if(flags != 0, "flags should be 0");
+
+  auto hip_wait_stream = get_stream(stream);
+  throw_invalid_resource_if(!hip_wait_stream, "stream is invalid");
+
+  throw_invalid_handle_if(!ev, "event is nullptr");
+  auto hip_event_cmd = std::dynamic_pointer_cast<event>(command_cache.get(ev));
+  throw_invalid_resource_if(!hip_event_cmd, "event is invalid");
 
-  throw std::runtime_error("Not implemented");
+  throw_if(!hip_event_cmd->is_recorded(), hipErrorStreamCaptureIsolation, "Event passed is not recorded");
+  auto hip_event_stream = hip_event_cmd->get_stream();
+
+  if (hip_wait_stream == hip_event_stream) {
+    hip_wait_stream->record_top_event(hip_event_cmd.get());
+  }
+  else {
+    auto wait_stream = hip_wait_stream.get();
```

Comment:
It is needed because in the next set of lines where record is called we std::move the shared_ptr hip_wait_stream so its no longer available for dereferencing